sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/datastore/DatastoreV1.java

```diff
@@ -359,19 +382,72 @@ static RunQueryRequest makeRequest(Query query, @Nullable String namespace) {
         Datastore datastore, QuerySplitter querySplitter, int numSplits) throws DatastoreException {
       // If namespace is set, include it in the split request so splits are calculated accordingly.
       PartitionId.Builder partitionBuilder = PartitionId.newBuilder();
-      if (namespace != null) {
+      if (!Strings.isNullOrEmpty(namespace)) {
         partitionBuilder.setNamespaceId(namespace);
       }
 
       return querySplitter.getSplits(query, partitionBuilder.build(), numSplits, datastore);
     }
 
     /**
+     * Translates a Cloud Datastore gql query string to {@link Query}.
+     *
+     * <p>Currently, the only way to translate a gql query string to a Query is to run the query
+     * against Cloud Datastore and extract the {@code Query} from the response. To prevent reading
+     * any data, we set the {@code LIMIT} to 0 but if the gql query already has a limit set, we
+     * catch the exception with {@code INVALID_ARGUMENT} error code and retry the translation
+     * without the zero limit.
+     *
+     * <p>Note: This may result in reading actual data from Cloud Datastore but the service has a
+     * cap on the number of entities returned for a single rpc request, so this should not be a
+     * problem in practice.
+     */
+    @VisibleForTesting
+    static Query translateGqlQueryWithLimitCheck(String gql, Datastore datastore,
+        String namespace) throws DatastoreException {
+      String gqlQueryWithZeroLimit = gql + " LIMIT 0";
+      try {
+        Query translatedQuery = translateGqlQuery(gqlQueryWithZeroLimit, datastore, namespace);
+        // Clear the limit that we set.
+        return translatedQuery.toBuilder().clearLimit().build();
+      } catch (DatastoreException e) {
+        // Note: There is no specific error code or message to detect if the query already has a
+        // limit, so we just check for INVALID_ARGUMENT and assume that that the query might have
+        // a limit already set.
+        if (e.getCode() == Code.INVALID_ARGUMENT) {
+          LOG.warn("Failed to translate Gql query '{}': {}", gqlQueryWithZeroLimit, e.getMessage());
+          LOG.warn("User query might have a limit already set, so trying without zero limit");
+          // Retry without the zero limit.
+          return translateGqlQuery(gql, datastore, namespace);
+        } else {
+          throw e;
+        }
+      }
+    }
+
+    /** Translates a gql query string to {@link Query}.*/
+    private static Query translateGqlQuery(String gql, Datastore datastore, String namespace)
+        throws DatastoreException {
+      GqlQuery gqlQuery = GqlQuery.newBuilder().setQueryString(gql)
+          .setAllowLiterals(true).build();
```

Comment:
Datastore team mentioned that as long as the query is from a trusted user, allowing literals is fine. Discussed with Eugene offline, we decided to rename the fields with 'literalsAllowed' prefix to be more explicit and also added a note in the documentation that user's need to trust the source of the query to avoid any security vulnerabilities.